File: src/mergeIntoMap/mergeIntoMap.ml
let mergeIntoMap original items getKey =
  let open Js in
  let open Array in
  let open Map in
  let changedItems =
    filter items ~f:(fun item ->
        match get original ~key:(getKey item) with
        | None -> true
        | Some originalItem -> not (Router.deepEqual item originalItem))
  in
  if length changedItems == 0 then original
  else
    let init = original |> entries |> Iterator.toArray |> fromArray in
    reduce changedItems ~init ~f:(fun map newItem ->
        map |> set ~key:(getKey newItem) ~value:newItem)

File: src/mergeIntoMap/mergeIntoMap.ts
import { deepEqual } from '@tanstack/react-router';

export function mergeIntoMap<K, V>(
  original: Map<K, V>,
  items: V[],
  getKey: (value: V) => K,
): Map<K, V> {
  const changedItems = items.filter(item => {
    const originalItem = original.get(getKey(item));
    if (originalItem == null) return true;

    return !deepEqual(item, originalItem);
  });

  if (changedItems.length === 0) return original;

  return changedItems.reduce(
    (map, newItem) => map.set(getKey(newItem), newItem),
    new Map(Array.from(original.entries())),
  );
}
